Add mk47 keyboard.

jonylee1986 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

jonylee1986 commented:
Hfdkb modify something from Suggested,tks!

waffle87 commented:
1. The changes I suggested on #19782 need to be applied to that pull request, not in a new one.
2. If you're adding multiple keyboards, each needs their own pull request.

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/config.h:
```diff
@@ -0,0 +1,130 @@
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION ROW2COL
+
+/* Hold ESC on start up to clear EEPROM and boot into bootloader mode */
+#define BOOTMAGIC_LITE_ROW 0
+#define BOOTMAGIC_LITE_COLUMN 0
```
Not needed, since this is the default
```suggestion
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,69 @@
+#include QMK_KEYBOARD_H
+
+enum custom_keycodes {
+    KC_MISSION_CONTROL = SAFE_RANGE,
```
```suggestion
    KC_MISSION_CONTROL = QK_USER,
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,69 @@
+};
+// clang-format on
+
+#ifdef RGB_MATRIX_ENABLE
+bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max) {
+    if (!rgb_matrix_indicators_advanced_user(led_min, led_max)) {
+        return false;
+    }
+    // caps lock cyan
+    if (host_keyboard_led_state().caps_lock) {
+        RGB_MATRIX_INDICATOR_SET_COLOR(12, 255, 255, 255);
+    }
+    return true;
+}
+#endif
```
```suggestion
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,69 @@
+};
+// clang-format on
+
+#ifdef RGB_MATRIX_ENABLE
+bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max) {
+    if (!rgb_matrix_indicators_advanced_user(led_min, led_max)) {
+        return false;
+    }
+    // caps lock cyan
+    if (host_keyboard_led_state().caps_lock) {
+        RGB_MATRIX_INDICATOR_SET_COLOR(12, 255, 255, 255);
+    }
+    return true;
+}
+#endif
```
```suggestion
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/matrix.c:
```diff
@@ -0,0 +1,183 @@
+    }
+}
+
+void matrix_scan_kb(void) {
```
```suggestion
void housekeeping_task_kb(void) {
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,170 @@
+};
+#endif
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
    if (!process_record_user(keycode, record)) {
        return false;
    }
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,170 @@
+            return true;  // Process all other keycodes normally
+    }
+}
+
```
```suggestion

#ifdef RGB_MATRIX_ENABLE
bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max) {
    if (!rgb_matrix_indicators_advanced_user(led_min, led_max)) {
        return false;
    }
    // caps lock cyan
    if (host_keyboard_led_state().caps_lock) {
        RGB_MATRIX_INDICATOR_SET_COLOR(12, 255, 255, 255);
    }
    return true;
}
#endif
```

jonylee1986 commented:
Please review again. @drashna

jonylee1986 commented:
> 

OK, I have corrected the title and content

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/matrix.c:
```diff
@@ -0,0 +1,199 @@
+/* Copyright (C) 2022 jonylee@hfd
```
All of this should be moved to `mk47.c` file.

jonylee1986 replied:
ok

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/config.h:
```diff
@@ -0,0 +1,121 @@
+#define I2C1_OPMODE OPMODE_I2C
+#define I2C1_CLOCK_SPEED 400000 /* 400000 */
+
+#define DRIVER_1_LED_TOTAL 47
+#define DRIVER_LED_TOTAL   DRIVER_1_LED_TOTAL
+#define RGB_MATRIX_LED_COUNT    DRIVER_1_LED_TOTAL
```
```suggestion
#define RGB_MATRIX_LED_COUNT 47
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,57 @@
+
+#include QMK_KEYBOARD_H
+
+enum custom_keycodes {
+    KC_MISSION_CONTROL = QK_USER,
+    KC_LAUNCHPAD
+};
```
These don't do anything?

jonylee1986 replied:
no,and  have deleted it

drashna replied:
delete looks to be unsuccessful.

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,57 @@
+
+#include QMK_KEYBOARD_H
+
+enum custom_keycodes {
+    KC_MISSION_CONTROL = SAFE_RANGE,
+    KC_LAUNCHPAD
+};
```
Above comment.

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/rules.mk:
```diff
@@ -0,0 +1,19 @@
+RGB_MATRIX_DRIVER = IS31FL3733
+EEPROM_DRIVER = wear_leveling
+WEAR_LEVELING_DRIVER = spi_flash
+
+SRC += matrix.c
```
```suggestion
```

waffle87 commented:
Please see the comment I marked unresolved. All references to `DRIVER_LED_TOTAL` should be removed.

jonylee1986 commented:
Removed all references to DRIVER_LED_TOTAL . @waffle87

fauxpark commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,339 @@
+#include "mk47.h"
+
+#ifdef RGB_MATRIX_ENABLE
+const is31_led PROGMEM g_is31_leds[47] = {
```
```suggestion
const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/config.h:
```diff
@@ -0,0 +1,119 @@
+//#define ENABLE_RGB_MATRIX_SOLID_SPLASH
+//#define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
+
+#define RGB_MATRIX_HUE_STEP 25
+#define RGB_MATRIX_SAT_STEP 25
+#define RGB_MATRIX_VAL_STEP 50
+#define RGB_MATRIX_SPD_STEP 52
```
As stated on multiple PRs of yours, these settings will cause likely undesirable behaviour for the end user.

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,332 @@
+#ifdef VIA_ENABLE
+        case KC_RESET: {
+            if (record->event.pressed) {
+#    include "via.h"
```
```suggestion
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,332 @@
+
+HSV hsv;
+
+void led_test(uint8_t color);
+void clear_eeprom(void);
+
```
```suggestion
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.h:
```diff
@@ -0,0 +1,35 @@
+
+#pragma once
+
+#include "quantum.h"
```
```suggestion
#include "quantum.h"

void led_test(uint8_t color);
void clear_eeprom(void);
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/readme.md:
```diff
@@ -0,0 +1,15 @@
+A customizable 40% keyboard.
+
+* Keyboard Maintainer: [jonylee@hfd](https://github.com/jonylee1986)
+* Hardware Supported: mk47
```
```suggestion
* Hardware Supported: mk47 PCB w/ WB32FQ95
```

waffle87 commented on keyboards/hfdkb/keyboard_sw/mk47/readme.md:
```diff
@@ -0,0 +1,15 @@
+Make example for this keyboard (after setting up your build environment):
+
+    make hfdkb/keyboard_sw/mk47:default
+
```
```suggestion

Flashing example for this keyboard:

    make hfdkb/keyboard_sw/mk47:default:flash
```

jonylee1986 commented:
Exm,is there anything else that needs to be modified? @waffle87

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,379 @@
+            }
+            return false;     
+#ifdef VIA_ENABLE
+        case KC_RESET: {
```
This isn't needed.  `QK_CLEAR_EEPROM` or `EE_CLR` also does this.

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,372 @@
+        }
+#endif
+     default:
+            return process_record_user(keycode, record);
```
`process_record_user` is already being checked, above. No need to repeat, or you may get double events. 
```suggestion
            return true;
```

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/mk47.c:
```diff
@@ -0,0 +1,372 @@
+            }
+            return false;     
+#ifdef VIA_ENABLE
+        case QK_KB: {
```
`QK_EEPROM_CLEAR` or `EE_CLR` already handles this case.

bbenne10 commented on keyboards/hfdkb/keyboard_sw/mk47/config.h:
I would like to see a `PRODUCT`, `MANUFACTURER`, and `DESCRIPTION` define herein, as recommended by the VIA guide. Additionally, I think we need a `PRODUCT_ID` and `VENDOR_ID` as well for VIA compatibility.

bbenne10 commented on keyboards/hfdkb/keyboard_sw/mk47/readme.md:
```diff
@@ -0,0 +1,18 @@
+
+    make hfdkb/keyboard_sw/mk47:default:flash
+
+**Reset Key**: Hold down the key located at *K01*, which programmed as *Esc* while plugging in the keyboard.
```
```suggestion
**Reset Key**: Hold down the key top-left key (default *Esc*) while plugging in the keyboard.
```

bbenne10 commented on keyboards/hfdkb/keyboard_sw/mk47/rules.mk:
```diff
@@ -0,0 +1,17 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
```
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic
```

This comment is out of sync, I think.

bbenne10 reviewed (commented):
I have tested this PR by cloning upstream and pulling a merge commit via: 
```console 
$ git pull origin refs/pull/19797/merge
```

This avoids the base being ~600 commits behind HEAD at the time of writing and allows to test against latest HEAD (which notably includes 2023q1 breaking changes that this PR currently lacks).

I copied the default keymap and tweaked to my liking and everything is working well.
I tested the default keymap only very briefly, as I found it miserable to work with (and the whole point of my looking into this PR is to avoid the default keymap). 
I have not yet tried the VIA support out, but will do so shortly.
However, I believe that the `PRODUCT_ID` and `VENDOR_ID` need inclusion before that will work.

jonylee1986 commented:
> I have tested this PR by cloning upstream and pulling a merge commit via:
> 
> ```
> $ git pull origin refs/pull/19797/merge
> ```
> 
> This avoids the base being ~600 commits behind HEAD at the time of writing and allows to test against latest HEAD (which notably includes 2023q1 breaking changes that this PR currently lacks).
> 
> I copied the default keymap and tweaked to my liking and everything is working well. I tested the default keymap only very briefly, as I found it miserable to work with (and the whole point of my looking into this PR is to avoid the default keymap). I have not yet tried the VIA support out, but will do so shortly. However, I believe that the `PRODUCT_ID` and `VENDOR_ID` need inclusion before that will work.

You can run the following command to compile a VIA supported bin file and burn it into your mk47.
```
make hfdkb/keyboard_sw/mk47:via
```
If you need a PID and VID, they are in info.json, and the MK47 already supports VIA. Even if you do not manually load a JSON file that supports VIA, the MK47 will still be automatically recognized by the VIA after you connect the MK47 to the computer.

drashna commented on keyboards/hfdkb/keyboard_sw/mk47/rules.mk:
```diff
@@ -0,0 +1,17 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = no       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = yes           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+ENCODER_ENABLE = no
+RGB_MATRIX_ENABLE = yes
```
These could mostly be moved to the info.json config.

zvecr reviewed (changes requested):
Readme is a bit sparse with availability info, but what does "keyboard_sw" reference to?

jonylee1986 commented:
> Readme is a bit sparse with availability info, but what does "keyboard_sw" reference to?

it is the name of a series，If don't like so many parent folders, I'll delete it🙏

waffle87 commented:
Can you provide a link to this series? Is "hfdkb" the vendor? The URL in `info.json` ([hfd.cn](https://hfd.cn)) does not go anywhere.

jonylee1986 commented:
> Can you provide a link to this series? Is "hfdkb" the vendor? The URL in `info.json` ([hfd.cn](https://hfd.cn)) does not go anywhere.

'mk47 ' is a product under the brand 'inland' and has not yet been launched. It will be launched on the 'micro center' platform in the future.
I have renamed the folder 'hfdkb' to 'inland' and modified readme.md and the URL in info.json. @waffle87

bbenne10 commented:
@jonylee1986: the Inland mk47 has absolutely launched at MicroCenter stores, at least in my area. My interest in this PR stems from having picked one up a few weeks ago. So far, it is quite a good board!

sigprof commented:
> I have renamed the folder 'hfdkb' to 'inland' and modified readme.md and the URL in info.json.

What does the `keyboard_sw` name component mean?  Maybe the keyboard name should be simply `inland/mk47` instead of the current `inland/keyboard_sw/mk47`?

sigprof commented on keyboards/inland/keyboard_sw/mk47/info.json:
```diff
@@ -0,0 +1,83 @@
+{
+    "keyboard_name": "MK47",
+    "manufacturer": "https://www.microcenter.com",
```
The `"manufacturer"` string will go to the `iManufacturer` USB string descriptor; it is unusual to put an URL there — in most cases a plain human readable name is used instead.  (Although apparently Windows does not display this field in any prominent place, in Linux you would see it in the `lsusb` output.)

There is the `"url"` field in JSON, but I'm not sure whether it is actually used anywhere at the moment.

qmkarst commented:
how were you able to flash to this board?

macexiv commented:
how were you able to flash to the inland mk47?

bbenne10 commented:
I flashed by holding the top left key on the board while booting, which placed the board in DFU.
QMK Toolkit can then flash it just fine.

macexiv commented:
> I flashed by holding the top left key on the board while booting, which placed the board in DFU. QMK Toolkit can then flash it just fine. 

I get the QMK  Toolkit message saying `USB Device Connected (NO DRIVER): WB Device in DFU Mode`

macexiv commented:
I installed and ran Zadig so I now have a driver but still can not flash. Do I need a special version of QMK Toolbox where this board is supported? If so, how do I go about installing that?

sigprof commented:
You need the [latest beta](https://github.com/qmk/qmk_toolbox/releases/tag/beta) release of QMK Toolbox for WB32 chips.

itarze commented:
@waffle87 Should the invalid label be removed ?

itarze commented on keyboards/inland/mk47/keymaps/inland/keymap.c:
```diff
@@ -0,0 +1,72 @@
+
+void housekeeping_task_user(void) {
+
+    if (pressed_time && (timer_elapsed(pressed_time) >= 1500)) {
```
```suggestion
    if (pressed_time && (timer_elapsed(pressed_time) >= 5000)) {
```

jonylee1986 commented:
@waffle87 please review，tks！

zvecr commented on keyboards/inland/mk47/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,47 @@
+        KC_ESC, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_BSPC,
+        KC_CAPS, KC_EXLM, KC_AT, KC_HASH, KC_DLR, KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_TRNS,
+        KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_UP, KC_SLSH,
+        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT),
+
+    [3] = LAYOUT(
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)
```
```suggestion
        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT)
```

zvecr commented on keyboards/inland/mk47/keymaps/inland/keymap.c:
```diff
@@ -0,0 +1,72 @@
+        KC_ESC, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_BSPC,
+        KC_CAPS, KC_EXLM, KC_AT, KC_HASH, KC_DLR, KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_TRNS,
+        KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_UP, KC_SLSH,
+        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT),
+
+    [3] = LAYOUT(
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)
```
```suggestion
        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT)
```

zvecr commented on keyboards/inland/mk47/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,47 @@
+        KC_ESC, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_BSPC,
+        KC_CAPS, KC_EXLM, KC_AT, KC_HASH, KC_DLR, KC_PERC, KC_CIRC, KC_AMPR, KC_ASTR, KC_LPRN, KC_RPRN, KC_TRNS,
+        KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_UP, KC_SLSH,
+        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT),
+
+    [3] = LAYOUT(
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
+        KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)
```
```suggestion
        KC_LCTL, KC_NO, KC_LALT, KC_DEL, KC_PGDN, KC_SPC, KC_PGUP, KC_NO, KC_LEFT, KC_DOWN, KC_RGHT)
```

zvecr commented on keyboards/inland/mk47/readme.md:
```diff
@@ -0,0 +1,19 @@
+
+    make inland/mk47:default:flash
+
+**Reset Key**: Hold down the key located at *K00*, which programmed as *Esc* while plugging in the keyboard.
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
```suggestion
See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

## Bootloader

Enter the bootloader in 2 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

zvecr commented on keyboards/inland/mk47/info.json:
```diff
@@ -0,0 +1,163 @@
+    "rgb_matrix": {
+        "driver": "IS31FL3733",
+        "max_brightness": 200,
+        "val_steps": 50,
+        "sat_steps": 65,
+        "hue_steps": 65,
+        "speed_steps": 65,
```
```suggestion
```

drashna commented on keyboards/inland/mk47/keymaps/inland/keymap.c:
```diff
@@ -0,0 +1,66 @@
+
+void housekeeping_task_user(void) {
+
+    if (pressed_time && (timer_elapsed(pressed_time) >= 5000)) {
```
deferred exec may work better for this, actually.

drashna commented on keyboards/inland/mk47/config.h:
```diff
@@ -0,0 +1,40 @@
+
+#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
+#define RGB_MATRIX_KEYPRESSES
+#define RGB_MATRIX_KEYRELEASES
```
Just a heads up, keyreleases takes precedence over keypress.
